File: dynamic/32/my.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
        
        # Using stack method
        # Stack stores the indices of characters
        stack = [-1]  # Initialize with -1 as base for empty stack calculations
        max_length = 0
        
        for i in range(len(s)):
            if s[i] == '(':
                # Push the index of opening parenthesis
                stack.append(i)
            else:  # s[i] == ')'
                # Pop the last opening parenthesis or base
                stack.pop()
                
                if not stack:
                    # If stack is empty, push current index as new base
                    stack.append(i)
                else:
                    # Calculate length of valid substring ending at current position
                    # Current index minus the last index in stack
                    max_length = max(max_length, i - stack[-1])
        
        return max_length
    
    def longestValidParentheses_dp(self, s):
        """
        Alternative solution using dynamic programming approach.
        """
        if not s:
            return 0
        
        n = len(s)
        # dp[i] = length of longest valid substring ending at i
        dp = [0] * n
        max_length = 0
        
        for i in range(1, n):
            if s[i] == ')':
                # Case 1: ()
                if s[i-1] == '(':
                    dp[i] = (dp[i-2] if i >= 2 else 0) + 2
                # Case 2: ))
                elif i - dp[i-1] > 0 and s[i - dp[i-1] - 1] == '(':
                    dp[i] = dp[i-1] + 2
                    # Add previous valid substring if it exists
                    if i - dp[i-1] - 2 >= 0:
                        dp[i] += dp[i - dp[i-1] - 2]
                
                max_length = max(max_length, dp[i])
        
        return max_length
        

File: stack/32/my.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int longestValidParentheses(char* s) {
    // Edge case: empty string
    if (s == NULL || *s == '\0') {
        return 0;
    }
    
    int length = strlen(s);
    int maxLength = 0;
    
    // Create a stack to store indices
    int* stack = (int*)malloc((length + 1) * sizeof(int));
    if (!stack) return 0; // Memory allocation check
    
    int top = 0;
    stack[top++] = -1; // Initialize with -1 as base for empty stack calculations
    
    for (int i = 0; i < length; i++) {
        if (s[i] == '(') {
            // Push the index of opening parenthesis
            stack[top++] = i;
        } else { // s[i] == ')'
            // Pop the last opening parenthesis or base
            top--;
            
            if (top == 0) {
                // If stack is empty, push current index as new base
                stack[top++] = i;
            } else {
                // Calculate length of valid substring ending at current position
                int currentLength = i - stack[top - 1];
                if (currentLength > maxLength) {
                    maxLength = currentLength;
                }
            }
        }
    }
    
    // Free allocated memory
    free(stack);
    
    return maxLength;
}